brainiak/matnorm_models/noise_covs.py

```diff
@@ -0,0 +1,268 @@
+import tensorflow as tf
+import numpy as np
+import abc
+from .helpers import define_scope, xx_t
+
+
+class NoiseCovBase:
+    """Base metaclass for noise covariances
+    """
+    __metaclass__ = abc.ABCMeta
+
+    loc = tf.placeholder(tf.float64, [None, None], name="positions")
+    mask = tf.placeholder(tf.float64, [None, None], name="cov_mask")
+
+    @abc.abstractmethod
+    def get_optimize_vars(self):
+        """ Returns a list of tf variables that need to get optimized to fit
+            this covariance
+        """
+        pass
+
+    @abc.abstractproperty
+    def logdet(self):
+        """ log|Sigma|
+        """
+        pass
+
+    @abc.abstractmethod
+    def Sigma_inv_x(self, X):
+        """Given this Sigma and some X, compute :math:`Sigma^{-1} * x`
+        """
+        pass
+
+    @define_scope
+    def Sigma_inv(self):
```

Comment:
I think mentioned above, but I'm doing this because sometimes Sigma_inv is trivial. Is it always the case that if Sigma_inv is trivial then Sigma_inv_x(I) is also trivial? Or are you saying that we should check inside Sigma_inv_x for X being identity?